atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/ClusterTopologyValidator.java

```diff
@@ -0,0 +1,203 @@
+import com.github.rholder.retry.Retryer;
+import com.github.rholder.retry.RetryerBuilder;
+import com.github.rholder.retry.StopStrategies;
+import com.github.rholder.retry.WaitStrategies;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Sets;
+import com.palantir.atlasdb.keyvalue.cassandra.pool.CassandraServer;
+import com.palantir.conjure.java.api.config.service.HumanReadableDuration;
+import com.palantir.logsafe.Preconditions;
+import com.palantir.logsafe.SafeArg;
+import com.palantir.logsafe.exceptions.SafeRuntimeException;
+import com.palantir.logsafe.logger.SafeLogger;
+import com.palantir.logsafe.logger.SafeLoggerFactory;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import one.util.streamex.EntryStream;
+import org.apache.thrift.transport.TTransportException;
+import org.immutables.value.Value;
+
+public final class ClusterTopologyValidator {
+    private static final SafeLogger log = SafeLoggerFactory.get(ClusterTopologyValidator.class);
+
+    public ClusterTopologyValidator() {}
+
+    public Set<CassandraServer> getNewHostsWithInconsistentTopologiesAndRetry(
+            Set<CassandraServer> newlyAddedHosts,
+            Map<CassandraServer, CassandraClientPoolingContainer> allHosts,
+            HumanReadableDuration waitTimeBetweenCalls,
+            HumanReadableDuration maxWaitTime) {
+        Retryer<Set<CassandraServer>> retryer = RetryerBuilder.<Set<CassandraServer>>newBuilder()
+                .retryIfResult(servers -> servers.size() == allHosts.size())
+                .retryIfException()
+                .withWaitStrategy(
+                        WaitStrategies.fixedWait(waitTimeBetweenCalls.toMilliseconds(), TimeUnit.MILLISECONDS))
+                .withStopStrategy(StopStrategies.stopAfterDelay(maxWaitTime.toMilliseconds(), TimeUnit.MILLISECONDS))
+                .build();
+        try {
+            return retryer.call(() -> getNewHostsWithInconsistentTopologies(newlyAddedHosts, allHosts));
+        } catch (ExecutionException | RetryException e) {
+            throw new SafeRuntimeException("Failed to obtain consistent view of hosts from cluster.", e);
+        }
+    }
+
+    public Set<CassandraServer> getNewHostsWithInconsistentTopologies(
+            Set<CassandraServer> newlyAddedHosts, Map<CassandraServer, CassandraClientPoolingContainer> allHosts) {
+
+        if (newlyAddedHosts.isEmpty()) {
+            return newlyAddedHosts;
+        }
+
+        Preconditions.checkArgument(
+                Sets.difference(newlyAddedHosts, allHosts.keySet()).isEmpty(),
+                "Newly added hosts must be a subset of all hosts.",
+                SafeArg.of("newlyAddedHosts", newlyAddedHosts),
+                SafeArg.of("allHosts", allHosts));
+
```

Comment:
High level bluf for below:

Basically, let's check and see if the current hosts (all hosts - new hosts) have formed a consensus on their topology (i.e., do they all see the same servers). If not, then we cannot validate the new hosts, so just report back all new servers as invalid (as we should remove them, as we cannot validate them!). 

Otherwise, if there is a consensus, but there are no hostIds (this is a valid state when no C* host supports the new endpoint OR we are starting up for the first time), then we will try and form a consensus with the new hosts. If no consensus, do not add new hosts, and if there is partial consensus, remove the hosts which do not match.